lib/utils/autocomplete.js

```diff
@@ -0,0 +1,46 @@
+'use strict';
+
+const homedir = require('os').homedir();
+const path = require('path');
+const md5 = require('md5');
+const readFileSync = require('./fs/readFileSync');
+const fileExistsSync = require('./fs/fileExistsSync');
+
+const Serverless = require('../Serverless');
+
+const tab = require('tabtab')({
+  name: 'serverless',
+});
+
+const getSuggestions = (cacheFilePath) => {
+  const cacheFile = readFileSync(cacheFilePath);
+
+  tab.on('serverless', (data, done) => {
+    done(null, Object.keys(cacheFile.commands));
+  });
+
+  Object.keys(cacheFile).forEach(command => {
+    tab.on(command, (data, done) => {
+      done(null, cacheFile.commands[command]);
+    });
+  });
+
+  tab.start();
+};
+
+const autocomplete = () => {
+  let servicePath = process.cwd();
+  if (!fileExistsSync(path.join(servicePath, 'serverless.yml'))) {
+    servicePath = 'x';
+  }
+  const cacheFilePath = `${homedir}/.serverless/cache/${md5(servicePath)}/autocomplete.json`;
+
+  if (!fileExistsSync(cacheFilePath)) {
```

Comment:
Resurfacing this comment since it hasn't been addressed but was hidden below the github changes.

> The cache is not guaranteed to be valid just because it exists. Users could change the plugins listed in servereless.yml which would change the outcome of autocomplete. To overcome this I would hash serverless.yml as a cache validation key and store it in the autocomplete.json. The autocomplete should then hash serverless.yml on every invocation and use that value to compare against the exisitng cache key in autocomplete.json. If the keys are different then the cache is invalid and serverless.init() should be called to regenerate it.